fetchHttpClient's getError() drops error.cause.code, breaking systemErrorRetryPolicy's retry classification for network failures

---

### Describe the bug

`fetchHttpClient.js`'s `getError()` only reads `error.code` when wrapping a failed
`fetch()` call into a `RestError`:

```js
function getError(e, request) {
  if (e && e?.name === "AbortError") {
    return e;
  } else {
    return new RestError(`Error sending request: ${e.message}`, {
      code: e?.code ?? RestError.REQUEST_SEND_ERROR,
      request
    });
  }
}
```

When Node's native `fetch()` (undici) fails at the TCP/TLS layer (connection reset,
timeout, DNS blip — no HTTP response received at all), it throws a **`TypeError`**
whose own `.code` is `undefined`; the actual errno (`ECONNRESET`, `ETIMEDOUT`,
`UND_ERR_SOCKET`, etc.) lives one level down, on **`error.cause.code`**:

```js
> const e = new TypeError('fetch failed', { cause: Object.assign(new Error(), { code: 'ECONNRESET' }) })
> e.code
undefined
> e.cause.code
'ECONNRESET'
```

Because `getError()` reads `e?.code` and never `e?.cause?.code`, every such failure
gets wrapped with the generic fallback code `RestError.REQUEST_SEND_ERROR` instead of
the real network error code.

This breaks retry classification downstream. `systemErrorRetryPolicy` →
`exponentialRetryStrategy.isSystemError()` only recognizes:

```js
function isSystemError(err) {
  return err.code === "ETIMEDOUT" || err.code === "ESOCKETTIMEDOUT" ||
    err.code === "ECONNREFUSED" || err.code === "ECONNRESET" ||
    err.code === "ENOENT" || err.code === "ENOTFOUND";
}
```

`REQUEST_SEND_ERROR` is not in that list, so `isSystemError` returns `false` for
**every** transient connection failure surfaced through the fetch-based HTTP client,
and `exponentialRetryStrategy.retry()` takes the `{ errorToThrow: responseError }`
branch — it rethrows immediately with **zero retries** on the very first hiccup,
even though the underlying condition (a plain TCP reset/timeout) is exactly the kind
of transient error the retry policy exists to absorb.

### Where this actually bit us

We hit this indirectly through `@azure/storage-blob`'s `BlockBlobClient.uploadFile()`
(used by `@actions/cache@6.1.0`'s `saveCacheV2`, which forces `useAzureSdk: true` for
every GitHub Actions Cache Service v2 upload). Large multi-chunk uploads
(`uploadConcurrency: 8`, `uploadChunkSize: 64 MiB`) fail intermittently with a bare
`TypeError: fetch failed` and zero retries — the more chunks/concurrent requests, the
higher the chance at least one hits a transient hiccup, and since any single failure
kills the whole upload with no retry, large archives fail far more often than small
ones. No diagnostic detail survives either, since `error.message` for this `TypeError`
is literally just the string `"fetch failed"` — the real cause is discarded at this
exact wrapping site.

### Expected behavior

`getError()` should check `e?.cause?.code` (falling back to `e?.code`, then the
generic `REQUEST_SEND_ERROR`) so `isSystemError()` can correctly classify — and retry
— genuine transient network failures surfaced through the fetch-based HTTP client:

```js
code: e?.code ?? e?.cause?.code ?? RestError.REQUEST_SEND_ERROR
```

### Reproduction

```js
const { RestError } = require("@typespec/ts-http-runtime");

// Simulates what Node's native fetch() throws on a TCP-layer failure
// (connection reset, timeout, DNS blip — no HTTP response received).
const fetchError = new TypeError("fetch failed", {
  cause: Object.assign(new Error("read ECONNRESET"), { code: "ECONNRESET" }),
});

console.log(fetchError.code);       // undefined  <- what getError() reads today
console.log(fetchError.cause.code); // 'ECONNRESET' <- where the real errno actually is
```

### Environment

- `@typespec/ts-http-runtime`: 0.3.5 (via `@azure/core-rest-pipeline@1.23.0` →
  `@azure/storage-blob@12.31.0` → `@actions/cache@6.1.0`)
- Node.js: 20/24 (GitHub Actions `ubuntu-latest` and self-hosted runners)
- Observed via: `rharkor/caching-for-turbo@v2.5.0` (a GitHub Action wrapping
  `@actions/cache` for Turborepo remote caching), where cache-save uploads fail
  intermittently (~70% of the time on large archives) with an undiagnosable
  `##[error]fetch failed` and no retry.

--- a/sdk/core/ts-http-runtime/CHANGELOG.md
+++ b/sdk/core/ts-http-runtime/CHANGELOG.md
@@ -9,6 +9,7 @@
 ### Bugs Fixed
 
 - `NodeHttpClient` request timeouts now remain active while buffered response bodies are being read. [Issue #39519](https://github.com/Azure/azure-sdk-for-js/issues/39519)
+- `FetchHttpClient` now surfaces the system error code (such as `ECONNRESET` or `ETIMEDOUT`) carried by the `cause` of a failed `fetch()` call on the resulting `RestError`, so that `systemErrorRetryPolicy` retries transient network failures instead of rethrowing them immediately. [Issue #39703](https://github.com/Azure/azure-sdk-for-js/issues/39703)
 
 ### Other Changes
 

--- a/sdk/core/ts-http-runtime/src/fetchHttpClient.ts
+++ b/sdk/core/ts-http-runtime/src/fetchHttpClient.ts
@@ -13,6 +13,7 @@ import { RestError } from "./restError.js";
 import { createHttpHeaders } from "./httpHeaders.js";
 import { isNodeReadableStream, isWebReadableStream } from "./util/typeGuards.js";
 import { arrayBufferViewToArrayBuffer } from "./util/arrayBuffer.js";
+import { isObject } from "./util/object.js";
 
 // The Fetch spec requires `duplex: "half"` when body is a ReadableStream,
 // but TypeScript's lib.dom.d.ts hasn't added it to RequestInit yet.
@@ -202,10 +203,21 @@ function getError(e: RestError, request: PipelineRequest): RestError {
     return e;
   } else {
     return new RestError(`Error sending request: ${e.message}`, {
-      code: e?.code ?? RestError.REQUEST_SEND_ERROR,
+      code: e?.code ?? getCauseCode(e) ?? RestError.REQUEST_SEND_ERROR,
       request,
     });
   }
+}
+
+/**
+ * Native fetch implementations such as undici in Node.js report network failures as a TypeError
+ * whose system error code (for example ECONNRESET, ETIMEDOUT or ENOTFOUND) is carried by the
+ * error's `cause` rather than by the error itself. Surfacing that code on the RestError lets
+ * retry policies classify the failure as a transient system error.
+ */
+function getCauseCode(e: Error): string | undefined {
+  const cause: unknown = e?.cause;
+  return isObject(cause) && typeof cause.code === "string" ? cause.code : undefined;
 }
 
 /**

--- a/sdk/core/ts-http-runtime/test/internal/browser/fetchHttpClient.spec.ts
+++ b/sdk/core/ts-http-runtime/test/internal/browser/fetchHttpClient.spec.ts
@@ -1,9 +1,16 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-import { describe, it, assert, vi, beforeEach, afterEach } from "vitest";
+import { describe, it, assert, expect, vi, beforeEach, afterEach } from "vitest";
 import { createFetchHttpClient } from "../../../src/fetchHttpClient.js";
-import { createPipelineRequest, createHttpHeaders, AbortError } from "../../../src/index.js";
+import {
+  createPipelineRequest,
+  createHttpHeaders,
+  AbortError,
+  RestError,
+  isRestError,
+} from "../../../src/index.js";
+import { systemErrorRetryPolicy } from "../../../src/policies/internal.js";
 import { png } from "./mocks/encodedPng.js";
 import { delay } from "../../../src/util/helpers.js";
 import { arrayBufferViewToArrayBuffer } from "../../../src/util/arrayBuffer.js";
@@ -512,6 +519,50 @@ describe("FetchHttpClient", function () {
     });
     const response = await client.sendRequest(request);
     assert.strictEqual(response.status, 200);
+  });
+
+  it("should surface the system error code carried by the cause of a failed fetch", async function () {
+    // Native fetch implementations such as undici throw a TypeError whose errno lives on the cause
+    const cause = Object.assign(new Error("read ECONNRESET"), { code: "ECONNRESET" });
+    vi.mocked(fetch).mockRejectedValue(new TypeError("fetch failed", { cause }));
+
+    const client = createFetchHttpClient();
+    const request = createPipelineRequest({ url: "https://localhost/reset" });
+    try {
+      await client.sendRequest(request);
+      assert.fail("Expected await to throw");
+    } catch (e: any) {
+      assert.isTrue(isRestError(e));
+      assert.strictEqual(e.code, "ECONNRESET");
+    }
+  });
+
+  it("should fall back to REQUEST_SEND_ERROR when neither the error nor its cause has a code", async function () {
+    vi.mocked(fetch).mockRejectedValue(new TypeError("fetch failed"));
+
+    const client = createFetchHttpClient();
+    const request = createPipelineRequest({ url: "https://localhost/reset" });
+    try {
+      await client.sendRequest(request);
+      assert.fail("Expected await to throw");
+    } catch (e: any) {
+      assert.isTrue(isRestError(e));
+      assert.strictEqual(e.code, RestError.REQUEST_SEND_ERROR);
+    }
+  });
+
+  it("should let systemErrorRetryPolicy retry a fetch failure caused by a system error", async function () {
+    const cause = Object.assign(new Error("read ECONNRESET"), { code: "ECONNRESET" });
+    vi.mocked(fetch)
+      .mockRejectedValueOnce(new TypeError("fetch failed", { cause }))
+      .mockResolvedValueOnce(createResponse(200));
+
+    const client = createFetchHttpClient();
+    const policy = systemErrorRetryPolicy({ retryDelayInMs: 1, maxRetryDelayInMs: 1 });
+    const request = createPipelineRequest({ url: "https://localhost/reset" });
+    const response = await policy.sendRequest(request, (req) => client.sendRequest(req));
+    assert.strictEqual(response.status, 200);
+    expect(fetch).toHaveBeenCalledTimes(2);
   });
 });
 

